applications/serial_lte_modem/src/slm_at_socket.c

```diff
@@ -1632,6 +1701,249 @@ static int handle_at_poll(enum at_parser_cmd_type cmd_type, struct at_parser *pa
+				afds[i].fd = INVALID_SOCKET;
+				continue;
+			}
+			afds[i].fd = socks[i].fd;
+
+			/* atomic_clear returns the value before clear. */
+			afds[i].events |= atomic_clear(&socks[i].async_poll.update_events);
+		}
+
+		ret = zsock_poll(afds, ARRAY_SIZE(afds), -1);
+		if (ret < 0) {
+			LOG_ERR("zsock_poll() failed: %d, exit thread...", -errno);
+			poll_ctx.poll_running = false;
+			break;
+		}
+		for (int i = 0; i < ARRAY_SIZE(afds); ++i) {
+			if (afds[i].fd == INVALID_SOCKET) {
+				continue;
+			}
+
+			if (afds[i].fd == poll_ctx.efd && afds[i].revents) {
+				LOG_DBG("EFD revents: 0x%04x", afds[i].revents);
+				if (afds[i].revents & ZSOCK_POLLIN) {
+					eventfd_t value;
+
+					eventfd_read(poll_ctx.efd, &value);
+					if (value == EFD_POLL) {
+						/* Restart polling. */
+						continue;
+					} else {
+						/* Exit thread. */
+						poll_ctx.poll_running = false;
+						break;
+					}
+				} else {
+					LOG_ERR("EFD revents: 0x%04x",
+						afds[i].revents);
+					break;
+				}
+			}
+
+			if (afds[i].revents) {
+				if (!in_datamode()) {
+					rsp_send_indicate("\r\n#XAPOLL: %d,0x%02x\r\n",
+							  afds[i].fd, afds[i].revents);
+				}
+
+				if (afds[i].revents &
+				    (ZSOCK_POLLERR | ZSOCK_POLLNVAL | ZSOCK_POLLHUP)) {
+					/* Remove socket from poll until closed. */
+					socks[i].async_poll.disable = true;
+				} else {
+					/* Remove POLLIN/POLLOUT from poll, until AT-operation. */
+					afds[i].events &= ~afds[i].revents;
+				}
+			}
+		}
+	}
+	LOG_DBG("Asynchronous polling thread stopped");
+}
+
+static int async_poll_start(void)
+{
+	if (poll_ctx.poll_running) {
+		LOG_DBG("Restart asynchronous polling");
+		return eventfd_write(poll_ctx.efd, EFD_POLL);
+	}
+
+	LOG_DBG("Start asynchronous polling");
+	if (poll_ctx.efd != INVALID_SOCKET) {
+		zsock_close(poll_ctx.efd);
+	}
+	poll_ctx.efd = eventfd(0, 0);
+	if (poll_ctx.efd < 0) {
+		LOG_ERR("eventfd() failed: %d", -errno);
+		return -errno;
+	}
+	poll_ctx.poll_running = true;
+	k_thread_create(&async_poll_thread_id, async_poll_thread_stack,
+			K_THREAD_STACK_SIZEOF(async_poll_thread_stack),
+			async_poll_thread, NULL, NULL, NULL,
+			K_LOWEST_APPLICATION_THREAD_PRIO, 0, K_NO_WAIT);
+	k_thread_name_set(&async_poll_thread_id, "Asynchronous polling");
+
+	return 0;
+}
+
+static int async_poll_stop(void)
+{
+	int err = 0;
+
+	if (poll_ctx.poll_running) {
+		LOG_DBG("Stop asynchronous polling");
+		err = eventfd_write(poll_ctx.efd, EFD_CLOSE);
+		if (err < 0) {
+			LOG_ERR("eventfd_write() failed: %d", -errno);
+		}
+		err = k_thread_join(&async_poll_thread_id, K_SECONDS(1));
+		if (err) {
+			LOG_WRN("k_thread_join() failed: %d", err);
+		}
+	}
+
+	if (poll_ctx.efd != INVALID_SOCKET) {
+		zsock_close(poll_ctx.efd);
+		poll_ctx.efd = INVALID_SOCKET;
+	}
+	poll_ctx.poll_events = 0;
+
+	return 0;
+}
+
+SLM_AT_CMD_CUSTOM(xapoll, "AT#XAPOLL", handle_at_async_poll);
+static int handle_at_async_poll(enum at_parser_cmd_type cmd_type, struct at_parser *parser,
```

Comment:
Should we name these as apoll instead of async_poll? I don't know if there are other handlers named differently to the AT command.
```suggestion
static int handle_at_apoll(enum at_parser_cmd_type cmd_type, struct at_parser *parser,
```

Reply:
In `libmodem` we have `nrf_modem_at_cmd_async`.

Reply:
Little bit more questionable here where people use the AT commands by typing them also. We don't use underscore in these and `_ASYNC` is getting a little "longish".